src/System.Private.ServiceModel/src/System/ServiceModel/Channels/ProducerConsumerStream.cs

```diff
@@ -0,0 +1,350 @@
+
+            if (_disposed)
+            {
+                return 0;
+            }
+
+            _buffer.TrySetResult(new WriteBufferWrapper(buffer, offset, count));
+            int totalBytesRead = 0;
+            do
+            {
+                int bytesRead = await _dataAvail.Task;
+                _dataAvail = new TaskCompletionSource<int>();
+                if (bytesRead == 0)
+                {
+                    break;
+                }
+
+                totalBytesRead += bytesRead;
+                count -= bytesRead;
+            } while (count > 0);
+
+            return totalBytesRead;
+        }
+
+        protected override void Dispose(bool disposing)
+        {
+            if (disposing && !_disposed)
+            {
+                _disposed = true;
+                _dataAvail.TrySetResult(0);
+            }
+
+            base.Dispose(disposing);
+        }
+
+        private struct WriteBufferWrapper : IEquatable<WriteBufferWrapper>
+        {
+            public static readonly WriteBufferWrapper EmptyContainer = new WriteBufferWrapper(null, -1, -1, true);
+
+            private byte[] _buffer;
+            private int _offset;
+            private int _count;
+            private readonly bool _readOnly;
+
+            public WriteBufferWrapper(byte[] buffer, int offset, int count) : this(buffer, offset, count, false) { }
+            private WriteBufferWrapper(byte[] buffer, int offset, int count, bool isReadOnly) : this()
+            {
+                _buffer = buffer;
+                _offset = offset;
+                _count = count;
+                _readOnly = isReadOnly;
+            }
+
+            public byte[] ByteBuffer
+            {
+                get { return _buffer; }
+                set
+                {
+                    if (_buffer == value)
+                    {
+                        return;
+                    }
+                    if (_readOnly)
+                    {
+                        throw new ArgumentException("value");
+                    }
+                    _buffer = value;
+                }
+            }
+
+            public int Offset
+            {
+                get { return _offset; }
+                set
+                {
+                    if (_offset == value)
+                    {
+                        return;
+                    }
+                    if (_readOnly)
+                    {
+                        throw new ArgumentException("value");
+                    }
+                    _offset = value;
+                }
+            }
+
+            public int Count
+            {
+                get { return _count; }
+                set
+                {
+                    if (_count == value)
+                    {
+                        return;
+                    }
+                    if (_readOnly)
+                    {
+                        throw new ArgumentException("value");
+                    }
```

Comment:
I've modified all the properties to be getters only. No code was using the set code path and the only getter that's being used is Count. It feels wrong to only expose count though, so I'm leaving all three getters even though only one of them is actually used.